lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntime.cpp

```diff
@@ -447,11 +447,45 @@ static bool HasReflectionInfo(ObjectFile *obj_file) {
   return hasReflectionSection;
 }
 
-SwiftLanguageRuntimeImpl::ReflectionContextInterface *
+SwiftLanguageRuntimeImpl::ThreadSafeReflectionContext 
 SwiftLanguageRuntimeImpl::GetReflectionContext() {
-  if (!m_initialized_reflection_ctx)
-    SetupReflection();
-  return m_reflection_ctx.get();
+  m_reflection_ctx_mutex.lock();
+  SetupReflection();
+  // SetupReflection can potentially fail.
+  if (m_initialized_reflection_ctx)
+    ProcessModulesToAdd();
+  return {m_reflection_ctx.get(), m_reflection_ctx_mutex};
+}
+
+void SwiftLanguageRuntimeImpl::ProcessModulesToAdd() {
+  // A snapshot of the modules to be processed. This is necessary because
+  // AddModuleToReflectionContext may recursively call into this function again.
+  ModuleList modules_to_add_snapshot;
+  {
+    std::lock_guard<std::recursive_mutex> lock(m_modules_to_add.GetMutex());
+    modules_to_add_snapshot = m_modules_to_add;
```

Comment:
```
modules_to_add_snapshot = std::move(m_modules_to_add);
```
https://en.cppreference.com/w/cpp/container/vector/vector
> 8) Move constructor. Constructs the container with the contents of other using move semantics. Allocator is obtained by move-construction from the allocator belonging to other. After the move, other is guaranteed to be [empty()](https://en.cppreference.com/w/cpp/container/vector/empty).

Reply:
(Sorry, I just realized this is a ModuleList, not a vector.) We could add a move constructor with similar guarantees to ModuleList though, and not extract the Mutex here.